spec/controllers/students_controller_spec.rb

```diff
@@ -1,104 +1,154 @@
+      end
     end
 
-    it 'does not accept files with invalid columns' do
-      post :upload, params: {
-        upload_file: fixture_file_upload(
-          'files/students/form_invalid_column.csv', 'text/csv')
-      }
+    describe '#delete_grace_period_deduction' do
+      it 'deletes an existing grace period deduction' do
+        grouping = create(:grouping_with_inviter, inviter: student)
+        deduction = create(:grace_period_deduction,
+                           membership: grouping.accepted_student_memberships.first,
+                           deduction: 1)
+        expect(student.grace_period_deductions.exists?).to be true
+        delete :delete_grace_period_deduction,
+               params: { id: student.id, deduction_id: deduction.id },
+               xhr: true
+        expect(grouping.grace_period_deductions.exists?).to be false
+      end
 
-      expect(response.status).to eq(302)
-      expect(flash[:error]).to_not be_empty
-      expect(response).to redirect_to action: 'index'
+      it 'raises a RecordNotFound error when given a grace period deduction that does not exist' do
+        expect do
+          delete :delete_grace_period_deduction,
+                 params: { id: student.id, deduction_id: 100 }
+        end.to raise_error(ActiveRecord::RecordNotFound)
+      end
 
-      expect(Student.where(last_name: 'Antheil')).to be_empty
-      expect(Student.where(user_name: 'c5bennet')).to be_empty
+      it 'raises a RecordNotFound error when given a grace period deduction for a different student' do
+        student2 = create(:student, grace_credits: 2)
+        grouping2 = create(:grouping_with_inviter, inviter: student2)
+        submission2 = create(:version_used_submission, grouping: grouping2)
+        create(:complete_result, submission: submission2)
+        deduction = create(:grace_period_deduction,
+                           membership: grouping2.accepted_student_memberships.first,
+                           deduction: 1)
+        expect do
+          delete :delete_grace_period_deduction,
+                 params: { id: student.id, deduction_id: deduction.id }
+        end.to raise_error(ActiveRecord::RecordNotFound)
+      end
     end
   end
 
-  describe '#delete_grace_period_deduction' do
-    it 'deletes an existing grace period deduction' do
-      grouping = create(:grouping_with_inviter, inviter: student)
-      deduction = create(:grace_period_deduction,
-                         membership: grouping.accepted_student_memberships.first,
-                         deduction: 1)
-      expect(student.grace_period_deductions.exists?).to be true
-      delete :delete_grace_period_deduction,
-             params: { id: student.id, deduction_id: deduction.id },
-             xhr: true
-      expect(grouping.grace_period_deductions.exists?).to be false
+  describe 'User is a Student' do
+    include ERB::Util
+    RSpec.shared_examples 'a student can' do
```

Comment:
You can probably come up with a more descriptive name than "a student can". You can discuss email notification settings, for example.